### ISSUE TITLE
twisted.test.test_log.FileObserverTestCase.test_timeFormatting fails in unusual timezones like Europe/Moscow

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5583]
|<img alt="marienz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| marienz reported|
|-|-|
|Trac ID|trac#5583|
|Type|defect|
|Created|2012-03-29 13:16:41Z|

test_timeFormatting does this:

```python
# There is no function in the time module which converts a UTC time
# tuple to a timestamp.
when = time.mktime((2001, 2, 3, 4, 5, 6, 7, 8, 0)) - time.timezone
```

This tries to get the unix timestamp for the utc datetime 2001-02-03 04:05:06 (which should not involve the local timezone at all). It does this by getting a unix timestamp for 2001-02-3 04:05:06 local time, then adjusting that by the local timezone offset. But time.mktime uses the timezone offset that was in effect at the local time passed to it (in 2001), while time.timezone is the current timezone offset. These two are not necessarily the same. Running trial with TZ=Europe/Moscow set results in:

```
twisted.trial.unittest.FailTest: not equal:
a = '2001-02-03 00:05:06-0500'
b = '2001-02-02 23:05:06-0500'
```

The comment hints at the solution: use calendar.timegm to directly get a unix timestamp from a UTC timetuple. Patch attached.

Attachments:

* [timeformatting.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/94c/94c8c1c0b3ac0c1440b7dfe3842b513115ce9a98/b73ebc3e7e7508ea38fa939e84675e435c6b21c3.patch) (879 bytes) - added by marienz on 2012-03-29 13:17:06Z - use calendar.timegm instead of time.mktime + time.timezone

<details><summary>Searchable metadata</summary>

```
trac-id__5583 5583
type__defect defect
reporter__marienz marienz
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1333027001000000 1333027001000000
changetime__1333891060000000 1333891060000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @itamarst|
|-|-|

Looks good, just needs 5583.misc newsfile and good to be merged (will do it myself in a bit if no one else gets to it).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set status to `closed`|
|-|-|

(In [34145]) Apply patch: use calendar.timegm instead of manual broken timezone math.

Author: marienz
Review: itamar
Fixes: [#5583](https://github.com/twisted/twisted/issues/5583)
